File: src/Domain/Activities/RegistrationExperience.cs
namespace Cohestra.Domain.Activities;

public static class RegistrationExperienceLayouts
{
    public const string Centered = "centered";
    public const string Split = "split";
    public const string Poster = "poster";
    public const string Immersive = "immersive";
    public const string Card = "card";

    public static readonly IReadOnlySet<string> All = new HashSet<string>(StringComparer.Ordinal)
    {
        Centered,
        Split,
        Poster,
        Immersive,
        Card,
    };
}

public static class RegistrationExperienceStyles
{
    public const string Modern = "modern";
    public const string Minimal = "minimal";
    public const string Editorial = "editorial";
    public const string Bold = "bold";
    public const string Soft = "soft";

    public static readonly IReadOnlySet<string> All = new HashSet<string>(StringComparer.Ordinal)
    {
        Modern,
        Minimal,
        Editorial,
        Bold,
        Soft,
    };
}

public static class RegistrationExperienceFlows
{
    public const string SinglePage = "single-page";
    public const string Sections = "sections";
    public const string StepByStep = "step-by-step";
    public const string Conversational = "conversational";

    public static readonly IReadOnlySet<string> All = new HashSet<string>(StringComparer.Ordinal)
    {
        SinglePage,
        Sections,
        StepByStep,
        Conversational,
    };
}

public static class RegistrationExperienceHeroDisplays
{
    public const string Cover = "cover";
    public const string Contain = "contain";
    public const string FullBleed = "full-bleed";
    public const string Split = "split";
    public const string Background = "background";
    public const string Hidden = "hidden";

    public static readonly IReadOnlySet<string> All = new HashSet<string>(StringComparer.Ordinal)
    {
        Cover,
        Contain,
        FullBleed,
        Split,
        Background,
        Hidden,
    };
}

public sealed class RegistrationExperience
{
    public string? Layout { get; set; }

    public string? Style { get; set; }

    public string? Flow { get; set; }

    public string? HeroDisplay { get; set; }
}

public sealed class ResolvedRegistrationExperience
{
    public required string Layout { get; init; }

    public required string Style { get; init; }

    public required string Flow { get; init; }

    public required string HeroDisplay { get; init; }
}

File: web/lib/registration-experience.ts
import type { RegistrationTheme, RegistrationThemePreset } from "@/lib/activities-api";

export type RegistrationExperienceLayout =
  | "centered"
  | "split"
  | "poster"
  | "immersive"
  | "card";

export type RegistrationExperienceStyle =
  | "modern"
  | "minimal"
  | "editorial"
  | "bold"
  | "soft";

export type RegistrationExperienceFlow =
  | "single-page"
  | "sections"
  | "step-by-step"
  | "conversational";

export type RegistrationExperienceHeroDisplay =
  | "cover"
  | "contain"
  | "full-bleed"
  | "split"
  | "background"
  | "hidden";

export type RegistrationExperienceConfig = {
  layout?: RegistrationExperienceLayout | null;
  style?: RegistrationExperienceStyle | null;
  flow?: RegistrationExperienceFlow | null;
  heroDisplay?: RegistrationExperienceHeroDisplay | null;
};

export type ResolvedRegistrationExperience = {
  layout: RegistrationExperienceLayout;
  style: RegistrationExperienceStyle;
  flow: RegistrationExperienceFlow;
  heroDisplay: RegistrationExperienceHeroDisplay;
};

export type RegistrationThemeWithExperience = RegistrationTheme & {
  experience?: RegistrationExperienceConfig | null;
  resolvedExperience?: ResolvedRegistrationExperience | null;
};

function defaultsFromPreset(preset: RegistrationThemePreset): ResolvedRegistrationExperience {
  switch (preset) {
    case "card":
      return {
        layout: "card",
        style: "modern",
        flow: "single-page",
        heroDisplay: "cover",
      };
    case "immersive":
      return {
        layout: "immersive",
        style: "modern",
        flow: "single-page",
        heroDisplay: "full-bleed",
      };
    case "compact":
      return {
        layout: "centered",
        style: "minimal",
        flow: "single-page",
        heroDisplay: "cover",
      };
    default:
      return {
        layout: "centered",
        style: "modern",
        flow: "single-page",
        heroDisplay: "cover",
      };
  }
}

const LAYOUTS = new Set<RegistrationExperienceLayout>([
  "centered",
  "split",
  "poster",
  "immersive",
  "card",
]);

/** Resolve composable experience from theme + legacy preset (mirrors API resolver). */
export function resolveRegistrationExperience(
  theme: RegistrationThemeWithExperience | null | undefined
): ResolvedRegistrationExperience {
  const preset = theme?.preset ?? "classic";
  const defaults = defaultsFromPreset(preset);
  const experience = theme?.experience ?? theme?.resolvedExperience ?? null;

  const pick = <T extends string>(
    value: T | null | undefined,
    allowed: Set<T>,
    fallback: T
  ): T => (value && allowed.has(value) ? value : fallback);

  return {
    layout: pick(
      experience?.layout as RegistrationExperienceLayout | null | undefined,
      LAYOUTS,
      defaults.layout
    ),
    style: pick(
      experience?.style as RegistrationExperienceStyle | null | undefined,
      new Set(["modern", "minimal", "editorial", "bold", "soft"]),
      defaults.style
    ),
    flow: pick(
      experience?.flow as RegistrationExperienceFlow | null | undefined,
      new Set(["single-page", "sections", "step-by-step", "conversational"]),
      defaults.flow
    ),
    heroDisplay: pick(
      experience?.heroDisplay as RegistrationExperienceHeroDisplay | null | undefined,
      new Set(["cover", "contain", "full-bleed", "split", "background", "hidden"]),
      defaults.heroDisplay
    ),
  };
}
